Integrate Fluxion with core's resource module

dongahn opened the pull request:
Note that #665 and #667 need to be merged first before this PR.

This PR integrates Fluxion with flux-core's resource module. 

- `sched-fluxion-resource`: perform the initial resource graph data store population when the first response of resource.acquire arrives. Update the resource graph whenever subsequence responses arrive. Its semantics and limitations are described in commit 1be7d51d2f05323e5df6caa87e2164eab5f890df

- `sched-fluxion-qmanager`: Send sched-fluxion-resource.notify streaming RPC to get synchronized with `sched-fluxion-resource` for resource acquision and state change events. Delay its handshaking protocol with job-manager until receiving the first response of `sched-fluxion-resource.notify`. Upon receiving each response, run schedule loop as it represents a scheduleable resource event. Stop the reactor and module exit when receiving the `ECANCELED` response of `sched-fluxion-resource.notify`, which occurs when sched-fluxion-resource is unloaded.

- Adjust test cases for newly created dependencies among flux-core's `resource`, `sched-simple` and fluxion's two service modules.

- Add cases to test correct handling of resource status changes (`t1011-dynstate-change.t`)

The high-level interaction diagram is available at https://github.com/flux-framework/flux-sched/issues/662#issuecomment-643496117.

dongahn commented:
Finally, this PR is ready for your review. Note that this PR depends on PR #693 though.

dongahn commented:
Well, one problem I found. When there are some administrative excluded ranks, fluxion wouldn't load:

resource.toml
```tomb
[resource]
exclude = "0"  
```

```console
ahn1@5efb1e8cd6ff:/usr/src$ flux start -s 4
2020-07-16T16:39:37.251704Z sched-fluxion-resource.err[0]: mark_now: traverser::mark: mark: could not find rank path in by_rank map.
2020-07-16T16:39:37.251745Z sched-fluxion-resource.err[0]: : Invalid argument
2020-07-16T16:39:37.251765Z sched-fluxion-resource.err[0]: init_resource_graph: mark (down)
2020-07-16T16:39:37.251779Z sched-fluxion-resource.err[0]: mod_main: can't initialize resource graph database
2020-07-16T16:39:37.251828Z sched-fluxion-resource.crit[0]: fatal error: Invalid argument
2020-07-16T16:39:37.253421Z broker.err[0]: rc1.0: flux-module: cmb.insmod: Invalid argument
2020-07-16T16:39:37.255632Z broker.err[0]: rc1.0: /bin/bash -c /etc/flux/rc1 Exited (rc=1) 0.5s
flux-start: 0 (pid 73670) exited with rc=1
```

This can be easily fixed by making the traverser's `mark()` method ignore unknown ranks.

```diff
index d29fc5cb..8ba9202a 100644
--- a/resource/traversers/dfu_impl_update.cpp
+++ b/resource/traversers/dfu_impl_update.cpp
@@ -613,12 +613,8 @@ int dfu_impl_t::mark (std::set<int64_t> &ranks,
         // Now iterate through subgraphs keyed by rank and
         // set status appropriately
         vit = m_graph_db->metadata.by_rank.find (rank);
-        if (vit == m_graph_db->metadata.by_rank.end ()) {
-            errno = EINVAL;
-            m_err_msg += __FUNCTION__;
-            m_err_msg += ": could not find rank path in by_rank map.\n";
-            return -1;
-        }
+        if (vit == m_graph_db->metadata.by_rank.end ())
+            continue;

         subtree_root = vit->second.front ();
         subtree_path = (*m_graph)[subtree_root].paths.at (dom);
```

@milroy: are you okay with this patch?

dongahn commented:
The other issue, if every rank is administratively excluded, fluxion load hangs. Need to look into this. Overall, it appears I need to add a bit more coverage against administrative exclusion.

dongahn commented:
> The other issue, if every rank is administratively excluded, fluxion load hangs. Need to look into this. Overall, it appears I need to add a bit more coverage against administrative exclusion.

This has been fixed. The code wasn't handling the special case correctly when all of the rank resources are excluded.

garlick commented on resource/modules/resource_match.cpp:
```diff
@@ -456,65 +457,296 @@ static int populate_resource_db_file (std::shared_ptr<resource_ctx_t> &ctx,
+        goto done;
+
+    json_object_foreach (o, k, v) {
+        struct idset *ids_tmp = NULL;
```
no need to set idset_tmp to NULL since you're about to clobber its value in the next line.

dongahn commented:
@garlick: thanks! I will address the issues as soon as practical. (Will be within today :-)

dongahn commented on resource/modules/resource_match.cpp:
```diff
@@ -456,65 +457,296 @@ static int populate_resource_db_file (std::shared_ptr<resource_ctx_t> &ctx,
+    if ( !(ids = idset_create (0, IDSET_FLAG_AUTOGROW)))
+        goto done;
+
+    json_object_foreach (o, k, v) {
+        struct idset *ids_tmp = NULL;
+        if ( !(ids_tmp = idset_decode (k)))
+            goto done;
+        unsigned int id = idset_first (ids_tmp);
+        while (id != IDSET_INVALID_ID) {
+            if (idset_set (ids, id) < 0)
+                goto done;
+            id = idset_next (ids_tmp, id);
+        }
+        idset_destroy (ids_tmp);
+    }
```
Good catch!

dongahn commented:
> You could just copy [this one](https://github.com/flux-framework/flux-core/blob/997d9b1f36ce13341a7bb4d7657d18c563c9cf23/src/modules/resource/rutil.c#L130)

Nice! Exactly same code sequence -- and unit test!

dongahn commented:
> In addition, compilation is currently (trivially) broken. To fix:

Yes, I noticed this yesterday and haven't had time to circle back. 

I will be working on it in the next few mins and as part of that I will look through other issues (merge conflict etc). Hopefully I can finish this up before @SteVwonder gets to this.

dongahn commented:
@garlick: Ok I addressed all of your comments, and cleaned up the merge issues and squashed those changes. The only thing that I didn't do was to pick the unit test as you suggested. The `resource_match.cpp` will need to be refactored first to introduce unit tests and we're not there yet. We have other unit tests and etc for the fluxion resource infrastructure proper (being driven by `resource-query`) but this code doesn't belong there. We'll ultimately need to refactor `resource_match.cpp`, and this will be considered at that time.

dongahn commented:
This should be ready to be reviewed by others as well...

I am wondering why codecov/patch is low though...

dongahn commented:
> I am wondering why codecov/patch is low though...

Looks to me like I need some tests to exercise the newly added `emit_json` for some writers at https://github.com/flux-framework/flux-sched/pull/694/commits/c45a54e15afa1da7bbf80b703fc6f3406f17503c

dongahn commented on resource/modules/resource_match.cpp:
```diff
@@ -465,65 +466,305 @@ static int populate_resource_db_file (std::shared_ptr<resource_ctx_t> &ctx,
+    unsigned int rank = idset_first (ids);
+    vtx_t v = boost::graph_traits<resource_graph_t>::null_vertex ();
+
+    if (db.metadata.roots.find ("containment") == db.metadata.roots.end ())
+        if ( (rc = grow (ctx, v, rank)) < 0)
+            goto done;
+
```
Yeah this would be better for readability. Because the if block would be consider as one AST node, the current code is still valid though. Sorry if it is too obvious.

dongahn commented on resource/modules/resource_match.cpp:
```diff
@@ -465,65 +466,305 @@ static int populate_resource_db_file (std::shared_ptr<resource_ctx_t> &ctx,
+            goto done;
+
     if (db.metadata.roots.find ("containment") == db.metadata.roots.end ()) {
+        rc = -1;
```
> When would this error path happen? Wouldn't a successful run of grow always result in at least the "cluster" vertex existing? If we are just being extra careful, that's fine; I just want to make sure my understanding is correct.

Yes for now this is just extra careful. Later, if the previous grow grows resource on a non-containment subsystem, we want to capture it.

dongahn replied:
OK. I will add a comment. Good thought.

dongahn commented on resource/modules/resource_match.cpp:
```diff
@@ -465,65 +466,305 @@ static int populate_resource_db_file (std::shared_ptr<resource_ctx_t> &ctx,
-        goto done;
+    if (rank == IDSET_INVALID_ID) {
+        // Grow cluster vertex and leave
+        if ( (rc = db.load ("", ctx->reader, rank)) < 0) {
```
When this condition only occurs when `rank == IDSET_INVALID_ID` and the `root` has not been registered for the `containment` hierarchy. So this is okay. 

If and when `grow` etc are factored out to be a part of public APIsm, more checks are needed as we will have no control over the caller. 

For now, this should be okay as these methods are static and we have full control over how other functions are using this within the same file scope.

dongahn commented on resource/modules/resource_match.cpp:
```diff
@@ -107,8 +119,36 @@ struct resource_ctx_t : public resource_interface_t {
+
+void msg_wrap_t::set_msg (const flux_msg_t *msg)
+{
+    m_msg = flux_msg_incref (msg);
```
Good catch! Yes you are right. This should be fixed. Thanks.

dongahn commented on resource/modules/resource_match.cpp:
```diff
@@ -1741,6 +1761,74 @@ static void get_property_request_cb (flux_t *h, flux_msg_handler_t *w,
+    if (ctx->notify_msgs.find (route) != ctx->notify_msgs.end ()) {
+        ctx->notify_msgs.erase (route);
+        flux_log (h, LOG_DEBUG, "%s: a notify request aborted", __FUNCTION__);
+    }
```
I initially had it that way and discovered that this callback was called quite frequently for normal interactions... So I decided not to log that condition. I forget exactly under what circumstances this callback was called (other than the notify streaming RPC ended with `qmanager` exits). 

I don't think I have any other `async` RPC other than `notify` that needs to be canceled on disconnect so this should be okay.

Now that I think about it, though, it would be useful to identify under what condition this happens. If I can find that condition again, I will open a ticket.

garlick replied:
It's not the greatest design, but you'll get a disconnect request any time a "client" that has sent you a request message disconnect.  It will appear to come from the original sender to allow you to track any state you have for that sender and get rid of it (e.g. terminate a streaming RPC).

dongahn replied:
Thanks @garlick. 

Yes, that was my understanding and I'm tracking disconnects from qmanager for tight sync.

I didn't think I need to log any other disconnects. Let me know if this should be logged.

dongahn commented:
Ok. I addressed all of the @SteVwonder's suggestions and forced a push.

If there are anything else that I need to address, please let me know as soon as possible. I do hope to merge this in to tag for our next flux-sched release. 

The team agreed to do our July code-release based on this and flux-core's v0.18 release (https://github.com/flux-framework/flux-core/pull/3073) and flux-accounting's v0.01.

As soon as this gets in, I will create a version of flux-core and flux-sched on our LC systems to get this to be used and tested by 1) a new mission-critical workflow through UQP/Themis (@ jameshcorbett) and 2) Perforce totalview (Peter Thompson -- @jdelsign: just realized I don't have Peter's GitHub ID).

milroy commented:
@dongahn I can provide a review later today.  If that's too late, please proceed without me.

dongahn commented:
> @dongahn I can provide a review later today. If that's too late, please proceed without me.

Please go ahead and take a look. I still have some time and I may add one more test code to increase the coverage as well.

dongahn commented:
@milroy: I added a couple more commits that will likely increase the code coverage for this PR: cd22223 and 6c6a96e. Could you quickly take a look at it? I believe this PR should go in if that's okay.

dongahn commented:
Rebased to pick up @garlick's PR #703.

dongahn commented:
@Mergifyio rebase.

dongahn commented:
Thanks every for your review!

milroy commented on resource/utilities/command.cpp:
```diff
@@ -334,12 +335,24 @@ int cmd_find (std::shared_ptr<resource_context_t> &ctx,
     }
-
-    out << o.str ();
+    if (o) {
```
It appears to be possible for `emit_json` to return 0 and this condition to evaluate as false.  I think this will happen if [`check_array_sizes`](https://github.com/flux-framework/flux-sched/blob/87fbe1bcd87010e831b07c1052f6617f728e86e7/resource/writers/match_writers.cpp#L168) returns < 1.

dongahn replied:
This is a bit convoluted. But the code should  be okay I think. There are two cases: 

1) `check_array_sizes` return <0, in which case the conditional statement (a few lines above) that checks the return code of `emit_json` will catch this error and return.

2) `check_array_sizes` return == 0, in which case `emit_json` will return 0 (because this is not an error) but  `o` will remain to be `nullptr`. So the if-branch at line #342 won't be executed.

milroy replied:
Yes, case 2 was the one I was considering.  I was thinking it wasn't desirable for the execution to fall through line 342, but after rereading the code I see this works as intended.  Thanks!

milroy commented:
Looks like I just missed the review window.  @dongahn if my comment merits further discussion I can open an issue.

dongahn commented:
> Looks like I just missed the review window. @dongahn if my comment merits further discussion I can open an issue.

My apology! I somehow thought you gave your approval. Looking back, it was @SteVwonder who gave his approval. Let me follow up on your posting in an inline comment.

SteVwonder commented:
Sorry for the confusion 🙈 .  I just saw that I hadn't formally approved and wanted to do so.
